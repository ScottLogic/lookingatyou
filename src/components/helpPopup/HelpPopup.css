.popup-content {
    padding: 15px !important;
    border-radius: 0.5em;
    font-size: 1.2vw;
}

.popup-content h1 {
    text-align: center;
}

.popup-content .icon {
    pointer-events: none;
    position: relative !important;
    top: 0em !important;
    left: 0em;
}
