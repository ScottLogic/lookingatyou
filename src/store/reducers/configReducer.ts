import {
    ConfigAction,
    ConfigSetAction,
    IAdvancedConfig,
    IAppConfig,
    IConfigState,
    IDetectionConfig,
    IModelConfig,
    ISetAdvancedConfigAction,
    ISetAppConfigAction,
    ISetDetectionConfigAction,
    ISetModelConfigAction,
} from '../actions/config/types';

export const initalModelConfig: IModelConfig = {
    architecture: 'MobileNetV1',
    outputStride: 16,
    inputResolution: 289,
    multiplier: 0.75,
};

export const initialDetectionConfig: IDetectionConfig = {
    flipHorizontal: false,
    maxDetections: 4,
    scoreThreshold: 0.6,
    nmsRadius: 20,
};

export const initialAppConfig: IAppConfig = {
    xSensitivity: 0.8,
    ySensitivity: 0.6,
    fps: 5,
    irisColor: '#55acee', // must be hex value, as this is passed to colour picker input
    toggleAdvanced: false,
    showHelp: true,
};

export const initialAdvancedConfig: IAdvancedConfig = {
    toggleDebug: false,
    modelConfig: initalModelConfig,
    detectionConfig: initialDetectionConfig,
    toggleReflection: true,
    reflectionOpacity: 0.2,
};

export const initialConfig: IConfigState = {
    advancedConfig: initialAdvancedConfig,
    appConfig: initialAppConfig,
};

const configActionMapping = {
    [ConfigSetAction.APP]: setAppConfig,
    [ConfigSetAction.ADVANCED]: setAdvancedConfig,
    [ConfigSetAction.MODEL]: setModelConfig,
    [ConfigSetAction.DETECTION]: setDetectionConfig,
    [ConfigSetAction.RESET]: resetConfig,
};

const configStore = (
    state: IConfigState = initialConfig,
    action: ConfigAction,
): IConfigState => {
    return configActionMapping.hasOwnProperty(action.type)
        ? configActionMapping[action.type](state, action)
        : state;
};

function setAppConfig(state: IConfigState, action: ConfigAction): IConfigState {
    return {
        ...state,
        appConfig: {
            ...state.appConfig,
            ...(action as ISetAppConfigAction).payload,
        },
    };
}

function setAdvancedConfig(
    state: IConfigState,
    action: ConfigAction,
): IConfigState {
    return {
        ...state,
        advancedConfig: {
            ...state.advancedConfig,
            ...(action as ISetAdvancedConfigAction).payload,
        },
    };
}

function setModelConfig(
    state: IConfigState,
    action: ConfigAction,
): IConfigState {
    return {
        ...state,
        advancedConfig: {
            ...state.advancedConfig,
            modelConfig: {
                ...state.advancedConfig.modelConfig,
                ...(action as ISetModelConfigAction).payload,
            },
        },
    };
}

function setDetectionConfig(
    state: IConfigState,
    action: ConfigAction,
): IConfigState {
    return {
        ...state,
        advancedConfig: {
            ...state.advancedConfig,
            detectionConfig: {
                ...state.advancedConfig.detectionConfig,
                ...(action as ISetDetectionConfigAction).payload,
            },
        },
    };
}

function resetConfig(): IConfigState {
    return {
        ...initialConfig,
        appConfig: { ...initialConfig.appConfig, showHelp: false },
    };
}

export default configStore;
