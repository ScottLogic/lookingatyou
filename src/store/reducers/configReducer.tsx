import { colours } from '../../AppConstants';
import {
    ConfigActionTypes,
    IConfigState,
    SET_CONFIG,
} from '../actions/config/types';

export const initialState: IConfigState = {
    config: {
        xSensitivity: 1,
        ySensitivity: 1,
        fps: 5,
        swapEyes: false,
        toggleDebug: false,
        irisColor: colours.irisColor,
    },
};

const configStore = (
    state: IConfigState = initialState,
    action: ConfigActionTypes,
): IConfigState => {
    switch (action.type) {
        case SET_CONFIG:
            return {
                ...state,
                ...(state.config = {
                    ...state.config,
                    ...action.payload.partialConfig,
                }),
            };
        default:
            return state;
    }
};

export default configStore;
