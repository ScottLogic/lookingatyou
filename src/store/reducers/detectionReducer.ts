import { eyelidPosition, maxNumTargetsToConsider } from '../../AppConstants';
import { IObjectDetector } from '../../models/objectDetection';
import {
    calculateColourMatch,
    IColour,
} from '../../utils/objectSelection/select';
import { ICoords } from '../../utils/types';
import {
    DetectionActionType,
    IDetectionState,
    ISetDetectionsActionPayload,
    SET_DETECTIONS,
    SET_IDLE_TARGET,
    SET_INTERVAL,
    SET_MODEL,
    SET_OPEN,
} from '../actions/detections/types';
import { getPreviousTarget } from '../selectors/detectionSelectors';

export const initialState: IDetectionState = {
    model: null,
    idleTarget: { left: { x: 0, y: 0 }, right: { x: 0, y: 0 } },
    detections: { left: [], right: [] },
    eyesOpenCoefficient: eyelidPosition.OPEN,
    detectionInterval: 0,
    history: [{ left: { x: 0, y: 0 }, right: { x: 0, y: 0 } }],
    colour: { r: 0, g: 0, b: 0 },
};

const detectionActionMapping = {
    [SET_MODEL]: setModel,
    [SET_INTERVAL]: setDetectionInterval,
    [SET_IDLE_TARGET]: setIdleTarget,
    [SET_DETECTIONS]: setDetections,
    [SET_OPEN]: setOpen,
};

const detectionStore = (
    state: IDetectionState = initialState,
    action: DetectionActionType,
): IDetectionState => {
    return action.type in detectionActionMapping
        ? detectionActionMapping[action.type](state, action)
        : state;
};

function setModel(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    return { ...state, model: action.payload as IObjectDetector };
}

function setDetectionInterval(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    return { ...state, detectionInterval: action.payload as number };
}

function setIdleTarget(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    return {
        ...state,
        idleTarget: {
            left: action.payload as ICoords,
            right: action.payload as ICoords,
        },
    };
}

function setDetections(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    const payload = action.payload as ISetDetectionsActionPayload;
    const newHistory = [...state.history];
    if (state.history.length >= maxNumTargetsToConsider) {
        newHistory.shift();
    }
    newHistory.push(payload.previousTarget);

    return {
        ...state,
        detections: payload.detections,
        history: newHistory,
    };
}

function setOpen(
    state: IDetectionState,
    action: DetectionActionType,
): IDetectionState {
    return { ...state, eyesOpenCoefficient: action.payload as number };
}

export default detectionStore;
