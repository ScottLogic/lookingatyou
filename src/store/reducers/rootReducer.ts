import { combineReducers } from 'redux';
import { IConfigState } from '../actions/config/types';
import { IDetectionState } from '../actions/detections/types';
import { IVideoState } from '../actions/video/types';
import configStore from './configReducer';
import detectionStore from './detectionReducer';
import videoStore from './videoReducer';

export default combineReducers({
    videoStore,
    configStore,
    detectionStore,
});

export interface IRootStore {
    videoStore: IVideoState;
    configStore: IConfigState;
    detectionStore: IDetectionState;
}
