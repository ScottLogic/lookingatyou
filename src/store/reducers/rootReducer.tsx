import { combineReducers } from 'redux';
import { IVideoState } from '../actions/video/types';
import videoStore from './videoReducer';

export default combineReducers({
    videoStore,
});

export interface IRootStore {
    videoStore: IVideoState;
}
