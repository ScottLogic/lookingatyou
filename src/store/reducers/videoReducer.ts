import {
    IVideo,
    IVideoState,
    SET_IMAGE_DATA,
    SET_VIDEO,
    SET_VIDEO_STREAMS,
    VideoActionTypes,
} from '../actions/video/types';

export const initialState: IVideoState = {
    videos: {},
    images: {},
};

const videoStore = (
    state: IVideoState = initialState,
    action: VideoActionTypes,
): IVideoState => {
    switch (action.type) {
        case SET_VIDEO_STREAMS:
            const newVideos: { [deviceId: string]: IVideo } = {};
            action.videos.map(
                (video: IVideo) => (newVideos[video.deviceId] = video),
            );
            return {
                ...state,
                videos: {
                    ...state.videos,
                    ...newVideos,
                },
            };
        case SET_VIDEO:
            const elementId = action.payload.deviceId;
            const updatedObject: { [key: string]: IVideo } = {};
            updatedObject[elementId] = {
                ...state.videos[elementId],
                ...action.payload,
            };
            return {
                ...state,
                videos: {
                    ...state.videos,
                    ...updatedObject,
                },
            };
        case SET_IMAGE_DATA:
            return {
                ...state,
                images: {
                    ...state.images,
                    ...action.images,
                },
            };
        default:
            return state;
    }
};

export default videoStore;
