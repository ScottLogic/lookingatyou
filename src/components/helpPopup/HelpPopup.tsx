import { Button } from '@material-ui/core';
import React from 'react';
import { connect } from 'react-redux';
import Popup from 'reactjs-popup';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { updateConfigAction } from '../../store/actions/config/actions';
import { ConfigSetAction } from '../../store/actions/config/types';
import { IRootStore } from '../../store/reducers/rootReducer';
import { getShowHelp } from '../../store/selectors/configSelectors';
import { IConfigMenuProps } from '../configMenu/ConfigMenu';
import './HelpPopup.css';

interface IHelpPopupProps {
    window: Window;
}

interface IHelpPopupMapStateToProps {
    showHelp: boolean;
}

interface IHelpPopupMapDispatchToProps {
    setShowHelpToFalse: () => void;
}

type HelpPopupProps = IHelpPopupProps &
    IHelpPopupMapStateToProps &
    IHelpPopupMapDispatchToProps;

function HelpPopup(props: HelpPopupProps) {
    return (
        <Popup
            open={props.showHelp}
            modal={true}
            onClose={props.setShowHelpToFalse}
            closeOnDocumentClick={false}
            closeOnEscape={false}
        >
            {close => (
                <>
                    <h1>Looking At You</h1>
                    This is an app designed to track users using a webcam. The
                    eyes will follow you around and react differently depending
                    on what they see. Hover over any of the menu items for more
                    information. Click the Help button to see this message
                    again.
                    <br />
                    <br />
                    <a
                        href="https://blog.scottlogic.com/2019/08/19/LookingAtYou.html"
                        target="_blank"
                        rel="noopener noreferrer"
                    >
                        Developed by interns at Scott Logic
                    </a>
                    <br />
                    <Button
                        variant="contained"
                        className="accept"
                        onClick={close}
                    >
                        Close
                    </Button>
                </>
            )}
        </Popup>
    );
}

const mapStateToProps = (state: IRootStore): IHelpPopupMapStateToProps => ({
    showHelp: getShowHelp(state),
});

const mapDispatchToProps = (
    dispatch: ThunkDispatch<IRootStore, void, Action>,
    ownProps: IConfigMenuProps,
) => ({
    setShowHelpToFalse: () =>
        dispatch(
            updateConfigAction(
                ConfigSetAction.APP,
                { showHelp: false },
                ownProps.window,
            ),
        ),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(HelpPopup);
