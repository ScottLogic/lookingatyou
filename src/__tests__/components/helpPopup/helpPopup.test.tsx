import { shallow } from 'enzyme';
import jsdom, { DOMWindow } from 'jsdom';
import React from 'react';
import { Provider } from 'react-redux';
import { Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import HelpPopup from '../../../components/helpPopup/HelpPopup';
import { initialConfig } from '../../../store/reducers/configReducer';

describe('HelpPopup', () => {
    let window: DOMWindow;
    let store: Store;
    beforeEach(() => {
        const mockStore = configureStore([thunk]);
        store = mockStore(initialConfig);
        window = new jsdom.JSDOM().window;
    });
    it('should render correctly', () => {
        const wrapper = shallow(
            <Provider store={store}>
                <HelpPopup window={window} />
            </Provider>,
        ).debug();
        expect(wrapper).toMatchSnapshot();
    });
});
