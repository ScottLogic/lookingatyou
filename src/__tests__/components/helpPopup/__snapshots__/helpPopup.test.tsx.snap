// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpPopup should render correctly 1`] = `
"<ContextProvider value={{...}}>
  <Connect(HelpPopup) window={{...}} />
</ContextProvider>"
`;
